File: .claude/hooks/pre-write-secret-detection.mjs
#!/usr/bin/env node
// @hookstack pre-write-secret-detection
// Bloque les écritures de fichiers contenant des secrets potentiels (PreToolUse Write|Edit)
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';

const SECRET_PATTERNS = [
  /(?:ANTHROPIC|OPENAI|CLAUDE|GEMINI|GROQ)_API_KEY\s*=\s*['"]?\S{20,}/i,
  /sk-(?:ant-|proj-)?[a-zA-Z0-9_-]{32,}/,
  /ghp_[a-zA-Z0-9]{36}/,
  /-----BEGIN (?:RSA |EC )?PRIVATE KEY/,
  /(?:password|passwd|secret|token)\s*=\s*['"][^'"]{6,}/i,
];

export function run(input) {
  const content =
    input.tool_input?.content ?? input.tool_input?.new_string ?? '';
  if (!content) return null;

  const match = SECRET_PATTERNS.find((p) => p.test(content));
  if (!match) return null;

  return {
    decision: 'block',
    reason:
      '[secret-detection] Potential secret in the content being written. Reference it via an environment variable or .env (gitignored) instead of hardcoding it.',
  };
}

/* v8 ignore next 5 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const input = JSON.parse(readFileSync(0, 'utf8'));
  const result = run(input);
  if (result) process.stdout.write(JSON.stringify(result));
}

File: .claude/hooks/pre-write-secret-detection.py
#!/usr/bin/env python3
# @hookstack pre-write-secret-detection
"""Blocks file writes containing potential secrets (PreToolUse Write|Edit)."""
import json
import re
import sys

SECRET_PATTERNS = [
    re.compile(r"(?:ANTHROPIC|OPENAI|CLAUDE|GEMINI|GROQ)_API_KEY\s*=\s*['\"]?\S{20,}", re.I),
    re.compile(r"sk-(?:ant-|proj-)?[a-zA-Z0-9_-]{32,}"),
    re.compile(r"ghp_[a-zA-Z0-9]{36}"),
    re.compile(r"-----BEGIN (?:RSA |EC )?PRIVATE KEY"),
    re.compile(r"(?:password|passwd|secret|token)\s*=\s*['\"][^'\"]{6,}", re.I),
]


def run(input_data):
    tool_input = input_data.get("tool_input") or {}
    content = tool_input.get("content") or tool_input.get("new_string") or ""
    if not content:
        return None

    for pattern in SECRET_PATTERNS:
        if pattern.search(content):
            return {
                "decision": "block",
                "reason": (
                    "[secret-detection] Potential secret in the content being written. Reference it via an "
                    "environment variable or .env (gitignored) instead of hardcoding it."
                ),
            }
    return None


if __name__ == "__main__":
    data = json.load(sys.stdin)
    result = run(data)
    if result:
        sys.stdout.write(json.dumps(result))
